File: erlang/seminars/tracer/src/vector.erl
-module(vector).

%%% A small module for 3D vector operations.

-export([smul/2, sub/2, add/2, dot/2, scale/2, normalize/1, norm/1, sq/1, cross/2]).

%% Scalar multiplication 

smul({X1,X2,X3}, S) ->
    {X1*S, X2*S, X3*S}.

%% Addition and subtarction

sub({X1,X2,X3},{Y1,Y2,Y3}) ->
    {X1-Y1,X2-Y2,X3-Y3}.

add({X1,X2,X3},{Y1,Y2,Y3}) ->
    {X1+Y1,X2+Y2,X3+Y3}.

%% Dot product

dot({X1,X2,X3},{Y1,Y2,Y3}) ->
    X1*Y1 + X2*Y2 + X3*Y3.


%% Scaling a vector to a specified lenth

scale(X, L) ->
    N = norm(X),
    %% this will crash if N == 0!
    smul(X, L/N).

%% Normalize by scaling to norm 1.

normalize(X) ->
    scale(X,1).

%% Square of a vector

sq(X) ->
    dot(X,X).

%% The norm (length) of a vector

norm({X1,X2,X3}) ->
    math:sqrt(X1*X1+X2*X2+X3*X3).

%% Cross product, used to find a vector that as ortogonal to both X
%% and Y

cross({X1,X2,X3},{Y1, Y2, Y3}) ->
    {X2*Y3-X3*Y2, X3*Y1-X1*Y3,  X1*Y2-X2*Y1}.


 

File: exercises/tracer/src/vector.ex
defmodule Vector do

  # Scalar multiplication.
  def smul({x, y, z}, s) do
    {x * s, y * s, z * s}
  end

  # Addition and subtarction.
  def sub({x1, y1, z1}, {x2, y2, z2}) do
    {x1 - x2, y1 - y2, z1 - z2}
  end

  def add({x1, y1, z1}, {x2, y2, z2}) do
    {x1 + x2, y1 + y2, z1 + z2}
  end

  # Dot product.
  def dot({x1, y1, z1}, {x2, y2, z2}) do
    x1 * x2 + y1 * y2 + z1 * z2
  end

  # Scaling a vector to a specified length.
  def scale(x, l) do
    n = norm(x)
    # this will crash if N == 0!
    smul(x, l / n)
  end

  # Normalize by scaling to norm 1.
  def normalize(x) do
    scale(x, 1)
  end

  # Square of a vector.
  def sq(x) do
    dot(x, x)
  end

  # The norm (length) of a vector.
  def norm({x, y, z}) do
    :math.sqrt(x * x + y * y + z * z)
  end

  # Cross product, used to find a vector that as ortogonal to 
  # both x and y.
  def cross({x1, y1, z1}, {x2, y2, z2}) do
    {y1 * z2 - z1 * y2, z1 * x2 - x1 * z2, x1 * y2 - y1 * x2}
  end

end
